### DISCUSSION TITLE
Maximum obstacles?

### DISCUSSION THREAD

**Discussion by quetzalsly** [recastnavigation/recastnavigation#666] [Q&A]
I have a question about maximum obstacles.

I found this hard coded in the code:

`static const int MAX_REQUESTS = 64;
ObstacleRequest m_reqs[MAX_REQUESTS];`

Is there any reason for this limit?
What is the difference between temp obstacles and convex volumes?

I want to navigate a level full of enemies, lets say there are 512 enemies, maybe they are not all present at the same time but I want to have them move and the pathfinding to avoid them using their radius when every frame the path changes because they move.

Any ideas for a good optimial solution?
Thanks.

**Comment by grahamboree** [comment]
Hi!  The limit here is fairly arbitrary, but a good idea for performance reasons.  Temporary obstacles are not meant to cover cases like you're describing where one agent is navigating around other agents.  That behavior is often referred to as "collision avoidance" and there are a number of approaches to solving that.  Temp obstacles are meant for stationary obstacles that should affect pathfinding but shouldn't be baked into the navmesh.  Doors are a great example of this.  Convex volumes (I assume you're referring to them in the navmesh bake stage) exist to define areas of the world that have different properties, such as different movement costs.  They're sometimes used to force large ares to be considered unwalkable.  Hope that helps.
